src/middlewared/middlewared/plugins/device_/vrrp_events.py

```diff
@@ -1,10 +1,196 @@
+from asyncio import sleep as asyncio_sleep
+from dataclasses import dataclass
+from collections import deque
+from logging import getLogger
+from math import floor
 from os import mkfifo
-from threading import Thread
-from time import sleep
+from threading import Thread, Event
+from time import sleep, time
 
+from middlewared.service import Service
 from middlewared.utils.prctl import set_name
 
-VRRP_THREAD = None
+LOGGER = getLogger('failover')  # so logs show up in /var/log/failover.log
+
+
+@dataclass
+class VrrpObjs:
+    fifo_thread = None
+    event_thread = None
+    event_queue = deque(maxlen=1)
+
+
+class VrrpThreadService(Service):
+    class Config:
+        cli_private = True
+        private = True
+
+    def pause_events(self):
+        if VrrpObjs.fifo_thread is not None and VrrpObjs.fifo_thread.is_alive():
+            VrrpObjs.fifo_thread.pause()
+        if VrrpObjs.event_thread is not None and VrrpObjs.event_thread.is_alive():
+            VrrpObjs.event_thread.pause()
+
+    def unpause_events(self):
+        if VrrpObjs.fifo_thread is not None and VrrpObjs.fifo_thread.is_alive():
+            VrrpObjs.fifo_thread.unpause()
+        if VrrpObjs.event_thread is not None and VrrpObjs.event_thread.is_alive():
+            VrrpObjs.event_thread.unpause()
+
+
+class VrrpEventThread(Thread):
+
+    def __init__(self, **kwargs):
+        super(VrrpEventThread, self).__init__()
+        self.middleware = kwargs.get('middleware')
+        self.event_queue = VrrpObjs.event_queue
+        self.shutdown_event = Event()
+        self.pause_event = Event()
+        self.grace_period = 0.5
+        self.user_provided_timeout = kwargs.get('timeout') or 2
+        self.max_wait = self.user_provided_timeout + self.grace_period
+        self.settle_time = (self.max_wait / 2) + self.grace_period
+        self.max_rapid_settle_time = 5
```

Comment:
It would probably be good at some point to have a file containing failover-related constants.